src/java.base/share/classes/jdk/internal/vm/vector/VectorSupport.java

```diff
@@ -185,82 +186,131 @@ public VectorMask(Object payload) {
     public abstract static class VectorPayloadMF {
         public abstract long multiFieldOffset();
 
-        private static VectorPayloadMF newInstanceFactory(int elemSize, int length)  {
+        @ForceInline
+        private static VectorPayloadMF newInstanceFactory(Class<?> elemType, int length) {
             VectorPayloadMF obj = null;
-            int vecSize = elemSize * length;
-            switch(vecSize) {
-                case 8: obj = new VectorPayloadMF64(); break;
-                case 16: obj = new VectorPayloadMF128(); break;
-                case 32: obj = new VectorPayloadMF256(); break;
-                case 64: obj = new VectorPayloadMF512(); break;
-                default: assert false : "Unhandled vector size";
+            if (elemType == byte.class) {
+                switch(length) {
+                    case  8: obj = new VectorPayloadMF64B(); break;
+                    case 16: obj = new VectorPayloadMF128B(); break;
+                    case 32: obj = new VectorPayloadMF256B(); break;
+                    case 64: obj = new VectorPayloadMF512B(); break;
+                    default: assert false : "Unhandled vector size";
```

Comment:
We could return the new object directly.
```suggestion
                    case  8: return new VectorPayloadMF64B();
                    case 16: return new VectorPayloadMF128B();
                    case 32: return new VectorPayloadMF256B();
                    case 64: return new VectorPayloadMF512B();
                    default: assert false : "Unhandled vector size";
```